source/adapters/opencl/program.cpp

```diff
@@ -316,20 +316,54 @@ UR_APIEXPORT ur_result_t UR_APICALL urProgramSetSpecializationConstants(
   CL_RETURN_ON_FAILURE(clGetProgramInfo(CLProg, CL_PROGRAM_CONTEXT, sizeof(Ctx),
                                         &Ctx, &RetSize));
 
-  cl_ext::clSetProgramSpecializationConstant_fn F = nullptr;
-  const ur_result_t URResult = cl_ext::getExtFuncFromContext<decltype(F)>(
-      Ctx, cl_ext::ExtFuncPtrCache->clSetProgramSpecializationConstantCache,
-      cl_ext::SetProgramSpecializationConstantName, &F);
+  std::unique_ptr<std::vector<cl_device_id>> DevicesInCtx;
+  cl_adapter::getDevicesFromContext(cl_adapter::cast<ur_context_handle_t>(Ctx),
+                                    DevicesInCtx);
 
-  if (URResult != UR_RESULT_SUCCESS) {
-    return URResult;
-  }
+  cl_platform_id CurPlatform;
+  clGetDeviceInfo((*DevicesInCtx)[0], CL_DEVICE_PLATFORM,
+                  sizeof(cl_platform_id), &CurPlatform, nullptr);
+
+  oclv::OpenCLVersion PlatVer;
+  cl_adapter::getPlatformVersion(CurPlatform, PlatVer);
+
+  bool UseExtensionLookup = false;
+  if (PlatVer < oclv::V2_2) {
+    UseExtensionLookup = true;
+  } else {
+    for (cl_device_id Dev : *DevicesInCtx) {
+      oclv::OpenCLVersion DevVer;
+
+      cl_adapter::getDeviceVersion(Dev, DevVer);
 
-  for (uint32_t i = 0; i < count; ++i) {
-    CL_RETURN_ON_FAILURE(F(CLProg, pSpecConstants[i].id, pSpecConstants[i].size,
-                           pSpecConstants[i].pValue));
+      if (DevVer < oclv::V2_2) {
+        UseExtensionLookup = true;
+        break;
+      }
+    }
   }
 
+  if (UseExtensionLookup == false) {
+    for (uint32_t i = 0; i < count; ++i) {
+      CL_RETURN_ON_FAILURE(clSetProgramSpecializationConstant(
+          CLProg, pSpecConstants[i].id, pSpecConstants[i].size,
+          pSpecConstants[i].pValue));
+    }
+  } else {
+    cl_ext::clSetProgramSpecializationConstant_fn F = nullptr;
```

Comment:
Could we can this `SetProgramSpecializationConstant` instead of `F`? Would be much clearer when its called lower down.

Reply:
Good idea, I've updated this now. I've done the same in #966 so probably a good idea to make the same change there too.